File: scripts/notices.py
from __future__ import annotations

import argparse
from pathlib import Path
import re
import sys
import tomllib
from typing import Any

from scripts.common import ROOT, ToolError, error_json, read_json, write_new


def _license_files(package: dict[str, Any]) -> list[Path]:
    manifest = Path(package['manifest_path'])
    if not manifest.is_absolute() or not manifest.is_file():
        raise ToolError('INVALID_CARGO_METADATA', 'Package manifest must exist at an absolute path')
    directory = manifest.parent.resolve()
    paths = set()
    declared = package.get('license_file')
    if declared is not None:
        if not isinstance(declared, str):
            raise ToolError('INVALID_LICENSE_PATH', 'Declared license file must be a path')
        license_path = directory / declared
        if license_path.is_symlink() or not license_path.is_file() or not license_path.resolve().is_relative_to(directory):
            raise ToolError('INVALID_LICENSE_PATH', 'Declared license file must remain inside the package')
        paths.add(license_path.resolve())
    inspected = 0
    for path in directory.iterdir():
        name = path.name.upper()
        if name.startswith(('LICENSE', 'LICENCE', 'COPYING', 'UNLICENSE', 'NOTICE')):
            if path.is_symlink():
                raise ToolError('INVALID_LICENSE_PATH', 'License resources cannot follow symbolic links')
            candidates = path.rglob('*') if path.is_dir() else [path]
            for candidate in candidates:
                inspected += 1
                if inspected > 512:
                    raise ToolError('LICENSE_RESOURCE_LIMIT', 'License directories exceed the resource traversal limit')
                if candidate.is_symlink() or not candidate.resolve().is_relative_to(directory):
                    raise ToolError('INVALID_LICENSE_PATH', 'License resources cannot follow symbolic links')
                if candidate.is_file():
                    paths.add(candidate.resolve())
    if not paths or len(paths) > 128:
        raise ToolError('MISSING_LICENSE_TEXT', 'Package must contain a bounded set of distributable license texts')
    return sorted(paths, key=lambda path: path.relative_to(directory).as_posix())


def generate_notices(metadata: dict[str, Any], lock_path: Path) -> str:
    if not isinstance(metadata, dict) or metadata.get('version') != 1 or not isinstance(metadata.get('packages'), list):
        raise ToolError('INVALID_CARGO_METADATA', 'Cargo metadata format version 1 is required')
    if not isinstance(metadata.get('resolve'), dict):
        raise ToolError('INCOMPLETE_CARGO_METADATA', 'Cargo metadata must include the resolved dependency graph')
    members = metadata.get('workspace_members')
    if not isinstance(members, list) or not all(isinstance(item, str) for item in members):
        raise ToolError('INVALID_CARGO_METADATA', 'Cargo workspace members are missing')
    lock = tomllib.loads(lock_path.read_text())
    locked = {(package['name'], package['version']) for package in lock['package']}
    packages = []
    seen = set()
    for package in metadata['packages']:
        if not isinstance(package, dict) or not isinstance(package.get('id'), str) or package['id'] in seen:
            raise ToolError('INVALID_CARGO_METADATA', 'Cargo package identities must be unique')
        seen.add(package['id'])
        if package['id'] in members:
            continue
        for key in ('name', 'version', 'manifest_path'):
            if not isinstance(package.get(key), str):
                raise ToolError('INVALID_CARGO_METADATA', 'Cargo package identity fields are missing')
        if (package['name'], package['version']) not in locked:
            raise ToolError('UNLOCKED_DEPENDENCY', 'Every dependency notice must match Cargo.lock')
        if re.fullmatch(r'[A-Za-z0-9_-]+', package['name']) is None:
            raise ToolError('INVALID_PACKAGE_NAME', 'Dependency names must be valid Cargo identifiers')
        license_expression = package.get('license')
        if license_expression is not None and (not isinstance(license_expression, str) or '\n' in license_expression):
            raise ToolError('INVALID_LICENSE_METADATA', 'License expression must be a single line')
        if not license_expression and not package.get('license_file'):
            raise ToolError('MISSING_LICENSE_METADATA', 'Dependency must declare its license or license file')
        packages.append(package)
    if not packages:
        raise ToolError('EMPTY_DEPENDENCY_NOTICES', 'Resolved dependency notices cannot be empty')
    output = ['# Third-party notices', '',
              'Generated from locked Cargo workspace metadata, including enabled build and test dependencies.',
              'This inventory is not a claim that every listed package is linked into the runtime binary.', '']
    total = 0
    for package in sorted(packages, key=lambda package: (package['name'], package['version'], package['id'])):
        output += [f"## {package['name']} {package['version']}", '',
                   f"Declared license: {package['license'] if package.get('license') else 'license-file'}", '']
        directory = Path(package['manifest_path']).parent.resolve()
        for path in _license_files(package):
            size = path.stat().st_size
            total += size
            if not 0 < size <= 2_097_152 or total > 16_777_216:
                raise ToolError('LICENSE_SIZE_LIMIT', 'License text exceeds the distribution limit')
            text = path.read_text(encoding='utf-8')
            output += [f'### {path.relative_to(directory).as_posix()}', '', text.rstrip(), '']
    return '\n'.join(output) + '\n'


def main() -> int:
    parser = argparse.ArgumentParser(description='Collect dependency license texts from locked Cargo metadata')
    parser.add_argument('--metadata', type=Path, required=True)
    parser.add_argument('--out', type=Path, required=True)
    arguments = parser.parse_args()
    try:
        output = generate_notices(read_json(arguments.metadata, 33_554_432), ROOT / 'Cargo.lock')
        write_new(arguments.out, output.encode(), 0o644)
        return 0
    except (ToolError, OSError, ValueError, KeyError) as error:
        sys.stderr.buffer.write(error_json(error))
        return 2


if __name__ == '__main__':
    raise SystemExit(main())

File: xtask/src/release/notices.rs
use crate::tooling::common::*;
use serde::Deserialize;
use serde_json::Value;
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Component, Path, PathBuf};

#[derive(Deserialize)]
struct Metadata {
    version: u64,
    packages: Vec<Package>,
    workspace_members: Vec<String>,
    resolve: Option<Value>,
}
#[derive(Deserialize)]
struct Package {
    id: String,
    name: String,
    version: String,
    manifest_path: PathBuf,
    license: Option<String>,
    license_file: Option<String>,
    source: Option<String>,
}

pub fn locked_packages(path: &Path) -> Result<BTreeSet<(String, String, Option<String>)>> {
    let bytes = read_bytes(path, 4_194_304)?;
    let mut packages = BTreeSet::new();
    let mut fields = BTreeMap::new();
    let mut in_package = false;
    for line in text(&bytes)?.lines().chain(std::iter::once("[end]")) {
        let line = line.trim();
        if line.starts_with('[') {
            if in_package {
                let name = fields.remove("name").ok_or_else(|| {
                    ToolError::new("INVALID_LOCKFILE", "Locked package has no name")
                })?;
                let version = fields.remove("version").ok_or_else(|| {
                    ToolError::new("INVALID_LOCKFILE", "Locked package has no version")
                })?;
                require(
                    packages.insert((name, version, fields.remove("source"))),
                    "INVALID_LOCKFILE",
                    "Locked package is duplicated",
                )?;
                fields.clear();
            }
            in_package = line == "[[package]]";
        } else if in_package && let Some((key, value)) = line.split_once('=') {
            let key = key.trim();
            if ["name", "version", "source"].contains(&key) {
                let value = parse_json(value.trim().as_bytes())?;
                require(
                    fields
                        .insert(key.to_owned(), string(&value)?.to_owned())
                        .is_none(),
                    "INVALID_LOCKFILE",
                    "Locked identity field is duplicated",
                )?;
            }
        }
    }
    require(
        !packages.is_empty(),
        "INVALID_LOCKFILE",
        "Lockfile has no package identities",
    )?;
    Ok(packages)
}

fn license_path(root: &Path, relative: &Path) -> Result<PathBuf> {
    let mut path = root.to_path_buf();
    require(
        !relative.as_os_str().is_empty(),
        "INVALID_LICENSE_PATH",
        "License resource path is empty",
    )?;
    for component in relative.components() {
        let Component::Normal(name) = component else {
            return Err(ToolError::new(
                "INVALID_LICENSE_PATH",
                "License resources must remain inside their package",
            ));
        };
        path.push(name);
        require(
            !fs::symlink_metadata(&path)?.file_type().is_symlink(),
            "INVALID_LICENSE_PATH",
            "License resources cannot follow symlinks",
        )?;
    }
    require(
        path.is_file() && path.canonicalize()?.starts_with(root),
        "INVALID_LICENSE_PATH",
        "License must be a contained regular file",
    )?;
    Ok(path)
}

fn license_files(package: &Package) -> Result<(PathBuf, Vec<PathBuf>)> {
    require(
        package.manifest_path.is_absolute()
            && fs::symlink_metadata(&package.manifest_path)?.is_file(),
        "INVALID_CARGO_METADATA",
        "Package manifest must be a regular file at an absolute path",
    )?;
    let directory = package
        .manifest_path
        .parent()
        .ok_or_else(|| {
            ToolError::new(
                "INVALID_CARGO_METADATA",
                "Package manifest has no directory",
            )
        })?
        .canonicalize()?;
    let mut paths = BTreeSet::new();
    if let Some(declared) = &package.license_file {
        paths.insert(license_path(&directory, Path::new(declared))?);
    }
    let mut inspected = 0usize;
    for item in fs::read_dir(&directory)? {
        let item = item?;
        let name = item.file_name().to_string_lossy().to_uppercase();
        if ["LICENSE", "LICENCE", "COPYING", "UNLICENSE", "NOTICE"]
            .iter()
            .any(|prefix| name.starts_with(prefix))
        {
            require(
                !item.file_type()?.is_symlink(),
                "INVALID_LICENSE_PATH",
                "License directories cannot follow symlinks",
            )?;
            let candidates = if item.file_type()?.is_dir() {
                list_files(&item.path(), 512)?
            } else {
                vec![item.path()]
            };
            inspected = inspected.checked_add(candidates.len()).ok_or_else(|| {
                ToolError::new("LICENSE_RESOURCE_LIMIT", "License resource count overflow")
            })?;
            require(
                inspected <= 512,
                "LICENSE_RESOURCE_LIMIT",
                "License resources exceed the traversal limit",
            )?;
            for path in candidates {
                require(
                    fs::symlink_metadata(&path)?.is_file(),
                    "INVALID_LICENSE_PATH",
                    "License must be a regular file",
                )?;
                paths.insert(path);
            }
        }
    }
    require(
        !paths.is_empty() && paths.len() <= 128,
        "MISSING_LICENSE_TEXT",
        "A bounded set of distributable license texts is required",
    )?;
    Ok((directory, paths.into_iter().collect()))
}

pub fn generate(metadata: Value, lock_path: &Path) -> Result<String> {
    let metadata: Metadata = decode(metadata)?;
    require(
        metadata.version == 1 && metadata.resolve.as_ref().is_some_and(Value::is_object),
        "INCOMPLETE_CARGO_METADATA",
        "Cargo metadata format version 1 with a resolved graph is required",
    )?;
    let locked = locked_packages(lock_path)?;
    let members: BTreeSet<_> = metadata.workspace_members.into_iter().collect();
    let mut seen = BTreeSet::new();
    let mut packages = Vec::new();
    for package in metadata.packages {
        require(
            seen.insert(package.id.clone()),
            "INVALID_CARGO_METADATA",
            "Package identities must be unique",
        )?;
        if members.contains(&package.id) {
            continue;
        }
        require(
            locked.contains(&(
                package.name.clone(),
                package.version.clone(),
                package.source.clone(),
            )),
            "UNLOCKED_DEPENDENCY",
            "Dependency identity and source must match Cargo.lock",
        )?;
        require(
            !package.name.is_empty()
                && package
                    .name
                    .bytes()
                    .all(|b| b.is_ascii_alphanumeric() || b"_-".contains(&b)),
            "INVALID_PACKAGE_NAME",
            "Dependency name is not a Cargo identifier",
        )?;
        checked_version(&package.version)?;
        require(
            package
                .license
                .as_ref()
                .is_none_or(|s| !s.contains(['\n', '\r', '\0']))
                && (package.license.as_ref().is_some_and(|s| !s.is_empty())
                    || package.license_file.is_some()),
            "MISSING_LICENSE_METADATA",
            "Dependency must declare a single-line license expression or license file",
        )?;
        packages.push(package);
    }
    require(
        !packages.is_empty(),
        "EMPTY_DEPENDENCY_NOTICES",
        "Resolved dependency notices cannot be empty",
    )?;
    packages.sort_by(|a, b| (&a.name, &a.version, &a.id).cmp(&(&b.name, &b.version, &b.id)));
    let mut output = String::from(
        "# Third-party notices\n\nGenerated from locked Cargo workspace metadata, including enabled build and test dependencies.\nThis inventory is not a claim that every listed package is linked into the runtime binary.\n\n",
    );
    let mut total = 0usize;
    for package in packages {
        output.push_str(&format!(
            "## {} {}\n\nDeclared license: {}\n\n",
            package.name,
            package.version,
            package
                .license
                .as_deref()
                .filter(|value| !value.is_empty())
                .unwrap_or("license-file")
        ));
        let (directory, files) = license_files(&package)?;
        for file in files {
            let bytes = read_bytes(&file, MAX_JSON_BYTES)?;
            total = total.checked_add(bytes.len()).ok_or_else(|| {
                ToolError::new("LICENSE_SIZE_LIMIT", "License byte count overflow")
            })?;
            require(
                !bytes.is_empty() && total <= 16_777_216,
                "LICENSE_SIZE_LIMIT",
                "License text exceeds the distribution limit",
            )?;
            let relative = file
                .strip_prefix(&directory)
                .map_err(|_| ToolError::new("INVALID_LICENSE_PATH", "License escaped its package"))?
                .to_string_lossy()
                .replace('\\', "/");
            output.push_str(&format!(
                "### {relative}\n\n{}\n\n",
                text(&bytes)?.trim_end()
            ));
        }
    }
    Ok(output)
}
